### DISCUSSION TITLE
How to approach "forked" monitors

### DISCUSSION THREAD

**Discussion by wsw70** [jamesoff/simplemonitor#1399] [Q&A]
I am considering writing a new Monitor that would look at specific entities in Home Assistant (an entity is, grosso modo, something that you switch or that you get some value of). I would filter the monitored entities by labels ("tags" - they are part of the attributes of an entity).

My problem is that I have **one** monitor that monitors **many** endpoints, retrieved dynamically. Each of the endpoints is a single, independent monitoring point (ie. an alert should be raised for each of them failing, individually).

I am at the "rough design" phase now and my main concern is how to manage this situation: one monitor and many monitored things.

I welcome any ideas, or suggestions in the discussion.

**Comment by jamesoff** [comment]
Been having a think about this and not sure what the best approach is yet. If it was just one monitor monitoring a number of things, and that one monitor would fail if any of the entities were bad (for whatever value of bad), that would be fine. The wrinkle is needing one monitor to manifest as multiple monitors at runtime so they can alert individually.

Currently my thinking is around having the "main" monitor (which you configure to point at HASS and with a tag to select entities with, or something) and that monitor "injects" additional monitors in to simplemonitor behind the scenes based on what it discovers. Something a bit like the [compund](https://github.com/jamesoff/simplemonitor/blob/develop/simplemonitor/Monitors/compound.py) monitor but in reverse maybe. It's likely to need changes to the main SimpleMonitor class to support it I think, something like having a `get_sub_monitors()` call on `Monitor` which the main loop invokes to discover the dynamic ones. These probably don't need to be full `Monitor` instances in their own right, but maybe a list of names which the main loop then calls `run_test()` on separately

```
for sub_monitor in this_monitor.get_sub_monitors():
     this_monitor.run_test(sub_monitor)
```

Similarly alerting and logging would need to handle that too, probably with some caching so that the list of sub monitors doesn't change between running tests and reporting outcomes (i.e. HASS finds a new matching entity during that time).

Bit rambly, but that's a bit of a brain dump to start things rolling :)

Out of interest, what's the use case? I'm a fan of HASS myself!

**Reply by wsw70** [reply]
Thanks for the extended "brain dump". I will probably start with a single Monitor that manages all of the entities and fails if any of them fail (this is what you mentioned at the very beginning).

> Out of interest, what's the use case? I'm a fan of HASS myself!

I am in love with this software, it made my life so much simpler! (my wife is terrified, though)

There are two main reasons:
- I have a very computer-dysfunctional family and I need to follow up with them plugging off stuff and not plugging it back on (or to the wrong outlet). And then they will say that a lamp does not work with the others...
- I would like to monitor some specific attributes of entities (temperature, battery level) to be warned when crossing some threshold

Two versions ago, they added labels to HA that makes it much easier to select what to monitor (possibly with labels such as `monitor-availability`, `monitor-battery`, ...) so I thought I would give that a try.

Starting simple (with one monitor) will be the way to go.

I will work on that as soon as
- I review your comments for the SSH monitor (and by review I mean "say thank you for pointing this out")
- I finish a dashboard for simplemonitor that replicates more or less your status page (that I use), but with Vue/Vite/Quasar and a more complete JSON dump file
